### ISSUE TITLE
the gradient of all parameters is None

### ISSUE THREAD

**Issue by nankepan** [hpcaitech/Open-Sora#283]
![image](https://github.com/hpcaitech/Open-Sora/assets/82510548/e4254cec-5400-4cc8-b1ac-ccca68c416ef)
Hi,
I print param.grad here and find that the gradient of all parameters is None.  Is this caused by using colorsalai? How can I obtain the gradient of parameters? Thank you.

**Comment by JThh** [comment]
It should only be `None` after `optimizer.zero_grad()`; `booster.backward` was doing `torch.optim.Optimizer.backward(loss)`. Would you mind printing the contents of `loss` to see if it is `NaN`?

**Comment by nankepan** [comment]
> It should only be `None` after `optimizer.zero_grad()`; `booster.backward` was doing `torch.optim.Optimizer.backward(loss)`. Would you mind printing the contents of `loss` to see if it is `NaN`?

Thanks for reply. `loss` is normal, but gradient is `None` before `optimizer.zero_grad()`, which is strange.

**Comment by zhengzangw** [comment]
This is because Colossalai manages the gradients, so you cannot directly access them by `param.grad`. @ver217 Could you please help with this?

**Comment by ver217** [comment]
Hi, gradients is managed in zero optimizer and `p.grad` is None. This is expected behavior. If you want to check grads manually, you can refer to https://github.com/hpcaitech/ColossalAI/blob/7f8b16635b42013b73e1cb1ffdebc07b4d71ac93/tests/test_zero/test_low_level/test_zero1_2.py#L164
Note that the grad is sharded and flat.

**Comment by 281LinChenjian** [comment]
I also need to extract p.grad for subsequent calculations. Is there any way to get p.grad correctly? I have read the above code but still don't know how to do it.

> Hi, gradients is managed in zero optimizer and `p.grad` is None. This is expected behavior. If you want to check grads manually, you can refer to https://github.com/hpcaitech/ColossalAI/blob/7f8b16635b42013b73e1cb1ffdebc07b4d71ac93/tests/test_zero/test_low_level/test_zero1_2.py#L164 Note that the grad is sharded and flat.

**Comment by zshyang** [comment]
same question here. wonder if someone has figure it out.

> I also need to extract p.grad for subsequent calculations. Is there any way to get p.grad correctly? I have read the above code but still don't know how to do it.
> 
> > Hi, gradients is managed in zero optimizer and `p.grad` is None. This is expected behavior. If you want to check grads manually, you can refer to https://github.com/hpcaitech/ColossalAI/blob/7f8b16635b42013b73e1cb1ffdebc07b4d71ac93/tests/test_zero/test_low_level/test_zero1_2.py#L164 Note that the grad is sharded and flat.

**Comment by zhengzangw** [comment]
Try: `optimizer.get_working_grad_by_param_id(id(LAYER_WEIGHT))`.
